File: platform-auth-server/src/main/java/com/webank/wecube/platform/auth/server/service/SubSystemManagementService.java
package com.webank.wecube.platform.auth.server.service;

import java.text.DateFormat;
import java.text.ParseException;
import java.text.SimpleDateFormat;
import java.util.ArrayList;
import java.util.Date;
import java.util.List;
import java.util.Set;

import javax.annotation.PostConstruct;

import org.apache.commons.lang3.StringUtils;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.security.authentication.BadCredentialsException;
import org.springframework.security.core.Authentication;
import org.springframework.stereotype.Service;

import com.webank.wecube.platform.auth.server.authentication.SubSystemAuthenticationToken;
import com.webank.wecube.platform.auth.server.common.ApplicationConstants;
import com.webank.wecube.platform.auth.server.common.AuthServerException;
import com.webank.wecube.platform.auth.server.config.AuthServerProperties;
import com.webank.wecube.platform.auth.server.dto.SimpleSubSystemDto;
import com.webank.wecube.platform.auth.server.dto.SubSystemTokenDto;
import com.webank.wecube.platform.auth.server.encryption.AsymmetricKeyPair;
import com.webank.wecube.platform.auth.server.encryption.EncryptionUtils;
import com.webank.wecube.platform.auth.server.entity.SysSubSystemEntity;
import com.webank.wecube.platform.auth.server.http.AuthenticationContextHolder;
import com.webank.wecube.platform.auth.server.http.AuthenticationContextHolder.AuthenticatedUser;
import com.webank.wecube.platform.auth.server.http.filter.DefaultJwtBuilder;
import com.webank.wecube.platform.auth.server.model.JwtToken;
import com.webank.wecube.platform.auth.server.model.SysSubSystemInfo;
import com.webank.wecube.platform.auth.server.repository.SubSystemRepository;

@Service("subSystemManagementService")
public class SubSystemManagementService {

    private static final Logger log = LoggerFactory.getLogger(SubSystemManagementService.class);

    private static final String DATE_PATTERN = "yyyyMMdd";

    private static final String MSG_BAD_CREDENTIAL = "Bad credential.";

    @Autowired
    private SubSystemInfoDataService subSystemInfoDataService;

    @Autowired
    private SubSystemRepository subSystemRepository;

    @Autowired
    private AuthServerProperties authServerProperties;

    private DefaultJwtBuilder jwtBuilder;

    @PostConstruct
    public void afterPropertiesSet() {
        jwtBuilder = new DefaultJwtBuilder(authServerProperties.getJwtToken());
    }

    /**
     * 
     * @param subSystemDto
     * @return
     */
    public SubSystemTokenDto registerSubSystemAccessToken(SubSystemTokenDto subSystemDto) {
        SubSystemTokenDto result = new SubSystemTokenDto();
        result.setSystemCode(subSystemDto.getSystemCode());

        if (!validateSubSystemTokenFields(subSystemDto)) {
            return result;
        }

        JwtToken accessToken = tryAuthenticateSubSystem(subSystemDto);
        if (accessToken == null) {
            return result;
        }

        result.setAccessToken(accessToken.getToken());
        result.setCreateDate(String.valueOf(System.currentTimeMillis()));
        result.setExpireDate(String.valueOf(accessToken.getExpiration()));

        return result;
    }

    /**
     * 
     * @param subSystemDto
     * @return
     */
    public SimpleSubSystemDto registerSubSystem(SimpleSubSystemDto subSystemDto) {

        validateStrictPermission();

        if (StringUtils.isBlank(subSystemDto.getSystemCode())) {
            throw new AuthServerException("3016", "Registering sub-system errors:system code cannot be blank.");
        }

        SysSubSystemEntity subSystem = subSystemRepository.findOneBySystemCode(subSystemDto.getSystemCode());

        if (subSystem == null) {
            subSystem = subSystemRepository.findOneBySystemName(subSystemDto.getName());
        }

        if (subSystem != null) {
            log.debug("such sub-system already exists,system code {}", subSystemDto.getSystemCode());
            return convertToSimpleSubSystemDto(subSystem);
        }
        
        log.info("About to create a new sub system:{}", subSystemDto);

        AsymmetricKeyPair keyPair = EncryptionUtils.initAsymmetricKeyPair();

        subSystem = new SysSubSystemEntity();
        subSystem.setCreatedBy(AuthenticationContextHolder.getCurrentUsername());
        subSystem.setDescription(subSystemDto.getDescription());
        subSystem.setApiKey(keyPair.getPrivateKey());
        subSystem.setPubApiKey(keyPair.getPublicKey());
        subSystem.setName(subSystemDto.getName());
        subSystem.setSystemCode(subSystemDto.getSystemCode());
        subSystem.setActive(true);
        subSystem.setBlocked(false);

        SysSubSystemEntity savedSubSystem = subSystemRepository.saveAndFlush(subSystem);

        return convertToSimpleSubSystemDto(savedSubSystem);
    }

    /**
     * 
     * @param systemCode
     * @return
     */
    public SimpleSubSystemDto retrieveSubSystemApikey(String systemCode) {
        SimpleSubSystemDto result = new SimpleSubSystemDto();
        if (StringUtils.isBlank(systemCode)) {
            return result;
        }

        SysSubSystemEntity subSystem = subSystemRepository.findOneBySystemCode(systemCode);
        if (subSystem == null) {
            return result;
        }

        result.setId(subSystem.getId());
        result.setActive(subSystem.isActive());
        result.setBlocked(subSystem.isBlocked());
        result.setDescription(subSystem.getDescription());
        result.setName(subSystem.getName());
        result.setSystemCode(subSystem.getSystemCode());
        result.setApikey(subSystem.getApiKey());

        return result;
    }

    /**
     * 
     * @param name
     * @return
     */
    public SimpleSubSystemDto retrieveSubSystemByName(String name) {
        validateStrictPermission();
        SysSubSystemEntity subSystem = subSystemRepository.findOneBySystemName(name);

        if (subSystem == null) {
            return null;
        }

        return convertToSimpleSubSystemDto(subSystem);
    }

    /**
     * 
     * @return
     */
    public List<SimpleSubSystemDto> retrieveAllSubSystems() {
        validateStrictPermission();
        List<SysSubSystemEntity> subSystems = subSystemRepository.findAll();

        List<SimpleSubSystemDto> result = new ArrayList<>();
        if (subSystems == null || subSystems.isEmpty()) {
            return result;
        }

        for (SysSubSystemEntity subSystem : subSystems) {
            SimpleSubSystemDto d = convertToSimpleSubSystemDto(subSystem);
            result.add(d);
        }
        return result;
    }

    protected void validateStrictPermission() {
        AuthenticatedUser currUser = AuthenticationContextHolder.getCurrentUser();
        if (currUser == null) {
            throw new AuthServerException("Lack of permission.");
        }

        Set<String> userRoles = currUser.getAuthorities();
        if (userRoles == null || userRoles.isEmpty()) {
            throw new AuthServerException("Lack of permission due to empty roles.");
        }

        if (!userRoles.contains(ApplicationConstants.Authority.SUBSYSTEM)) {
            throw new AuthServerException("Lack of permission due to no sub-system authority.");
        }

        if (ApplicationConstants.SubSystemName.SYS_PLATFORM.equalsIgnoreCase(currUser.getUsername())) {
            return;
        }

        log.info("current username is : {}", currUser.getUsername());
        throw new AuthServerException("Lack of permission due to not platform identity.");
    }

    private JwtToken tryAuthenticateSubSystem(SubSystemTokenDto subSystem) {
        String systemCode = subSystem.getSystemCode();
        SysSubSystemInfo subSystemInfo = subSystemInfoDataService.retrieveSysSubSystemInfoWithSystemCode(systemCode);
        if (subSystemInfo == null) {
            String msg = String.format("Sub system %s does not exist.", systemCode);
            throw new BadCredentialsException(msg);
        }

        if (subSystemInfo.getBlocked() == true) {
            throw new BadCredentialsException(String.format("Sub system %s is blocked.", systemCode));
        }

        if (subSystemInfo.getActive() == false) {
            throw new BadCredentialsException(String.format("Sub system %s is inactive.", systemCode));
        }

        return doAuthenticateSubSystem(subSystem, subSystemInfo);
    }

    private Authentication tryAuthenticate(SubSystemTokenDto subSystem, SysSubSystemInfo subSystemInfo) {
        Date currTime = new Date();
        String currTimeStr = formatDate(currTime);
        if (!currTimeStr.equals(subSystem.getCreateDate())) {
            throw new BadCredentialsException(MSG_BAD_CREDENTIAL);
        }

        if (subSystem.getNonce().length() != 10) {
            throw new BadCredentialsException(MSG_BAD_CREDENTIAL);
        }

        int len = Integer.parseInt(subSystem.getNonce().substring(8));
        if (len != subSystem.getSystemCode().length()) {
            throw new BadCredentialsException(MSG_BAD_CREDENTIAL);
        }

        return createSuccessAuthentication(subSystemInfo, subSystem);
    }

    protected Authentication createSuccessAuthentication(SysSubSystemInfo retrievedSubSystemInfo,
            SubSystemTokenDto subSystem) {

        SubSystemAuthenticationToken returnAuthToken = new SubSystemAuthenticationToken(subSystem.getSystemCode(),
                subSystem.getSystemCode(), subSystem.getNonce(), retrievedSubSystemInfo.getAuthorities());

        return returnAuthToken;

    }

    private JwtToken doAuthenticateSubSystem(SubSystemTokenDto subSystem, SysSubSystemInfo subSystemInfo) {
        Date expireTime = tryCalculateExpireTime(subSystem);
        Authentication auth = tryAuthenticate(subSystem, subSystemInfo);
        return jwtBuilder.buildAccessToken(auth, expireTime);
    }

    private Date tryCalculateExpireTime(SubSystemTokenDto subSystem) {
        Date expireDate = parseDate(subSystem.getExpireDate());
        return expireDate;
    }

    private Date parseDate(String dateTimeStr) {
        DateFormat df = new SimpleDateFormat(DATE_PATTERN);
        try {
            return df.parse(dateTimeStr);
        } catch (ParseException e) {
            throw new RuntimeException("Bad date format.");
        }
    }

    private String formatDate(Date dateTime) {
        DateFormat df = new SimpleDateFormat(DATE_PATTERN);
        return df.format(dateTime);
    }

    private boolean validateSubSystemTokenFields(SubSystemTokenDto dto) {
        if (dto == null) {
            return false;
        }

        if (StringUtils.isBlank(dto.getSystemCode())) {
            return false;
        }

        if (StringUtils.isBlank(dto.getNonce())) {
            return false;
        }

        if (StringUtils.isBlank(dto.getCreateDate())) {
            return false;
        }

        if (StringUtils.isBlank(dto.getExpireDate())) {
            return false;
        }

        return true;
    }

    private SimpleSubSystemDto convertToSimpleSubSystemDto(SysSubSystemEntity subSystem) {
        SimpleSubSystemDto dto = new SimpleSubSystemDto();
        dto.setId(subSystem.getId());
        dto.setActive(subSystem.isActive());
        dto.setBlocked(subSystem.isBlocked());
        dto.setDescription(subSystem.getDescription());
        dto.setName(subSystem.getName());
        dto.setSystemCode(subSystem.getSystemCode());
        dto.setApikey(subSystem.getApiKey());
        dto.setPubKey(subSystem.getPubApiKey());

        return dto;
    }
}

File: platform-auth-server/service/subsystem_management_service.go
package service

import (
	"errors"
	"fmt"
	"github.com/WeBankPartners/wecube-platform/platform-auth-server/common/constant"
	"github.com/WeBankPartners/wecube-platform/platform-auth-server/common/exterror"
	"github.com/WeBankPartners/wecube-platform/platform-auth-server/common/log"
	"github.com/WeBankPartners/wecube-platform/platform-auth-server/common/utils"
	"github.com/WeBankPartners/wecube-platform/platform-auth-server/model"
	"github.com/WeBankPartners/wecube-platform/platform-auth-server/service/db"
	"github.com/WeBankPartners/wecube-platform/platform-auth-server/service/jwt"
	"go.uber.org/zap"
	"strconv"
	"time"
)

const (
	MsgBadCredential = "Bad credential."
)

var SubSystemManagementServiceInstance SubSystemManagementService

type SubSystemManagementService struct {
}

func (SubSystemManagementService) RegisterSubSystemAccessToken(subSystemDto *model.SubSystemTokenDto) (*model.SubSystemTokenDto, error) {
	result := &model.SubSystemTokenDto{
		SystemCode: subSystemDto.SystemCode,
	}

	if !validateSubSystemTokenFields(subSystemDto) {
		return result, nil
	}

	accessToken, err := tryAuthenticateSubSystem(subSystemDto)
	if err != nil {
		return nil, err
	}
	if accessToken == nil {
		return result, nil
	}

	result.AccessToken = accessToken.Token
	result.CreateDate = strconv.Itoa(int(time.Now().UTC().Unix()))
	result.ExpireDate = accessToken.Expiration

	return result, nil
}

func validateSubSystemTokenFields(dto *model.SubSystemTokenDto) bool {
	if dto == nil {
		return false
	}

	if len(dto.SystemCode) == 0 {
		return false
	}

	if len(dto.Nonce) == 0 {
		return false
	}

	if len(dto.CreateDate) == 0 {
		return false
	}

	if len(dto.ExpireDate) == 0 {
		return false
	}

	return true
}

func tryAuthenticateSubSystem(subSystem *model.SubSystemTokenDto) (*model.JwtTokenDto, error) {
	systemCode := subSystem.SystemCode
	subSystemInfo, err := SubSystemInfoDataServiceImplInstance.retrieveSysSubSystemInfoWithSystemCode(systemCode)
	if err != nil {
		return nil, err
	}
	if subSystemInfo == nil {
		msg := fmt.Sprintf("Sub system %s does not exist.", systemCode)
		return nil, exterror.NewBadCredentialsError(msg)
	}

	if subSystemInfo.Blocked {
		return nil, exterror.NewBadCredentialsError(fmt.Sprintf("Sub system %s is blocked.", systemCode))
	}

	if !subSystemInfo.Active {
		return nil, exterror.NewBadCredentialsError(fmt.Sprintf("Sub system %s is inactive.", systemCode))
	}

	jwtToken, err := doAuthenticateSubSystem(subSystem, subSystemInfo)
	return jwtToken, err
}

func doAuthenticateSubSystem(subSystem *model.SubSystemTokenDto, subSystemInfo *model.SysSubSystemInfo) (*model.JwtTokenDto, error) {
	expireTime, err := tryCalculateExpireTime(subSystem)
	if err != nil {
		return nil, err
	}
	auth, err := tryAuthenticate(subSystem, subSystemInfo)
	if err != nil {
		return nil, err
	}

	jwtToken, err := jwt.BuildAccessToken(auth.Principal, auth.Authorities, constant.AuthoritySubsystem, *expireTime)
	if err != nil {
		return nil, err
	}

	return jwtToken, nil
}

func tryCalculateExpireTime(subSystem *model.SubSystemTokenDto) (*time.Time, error) {
	expireDate, err := parseDate(subSystem.ExpireDate)
	return expireDate, err
}

func parseDate(dateString string) (*time.Time, error) {
	layout := "20060102" //yyyyMMdd

	// Parse the input date string into a time object
	parsedTime, err := time.ParseInLocation(layout, dateString, time.Local)
	if err != nil {
		return nil, err
	}
	return &parsedTime, nil
}

func tryAuthenticate(subSystem *model.SubSystemTokenDto, subSystemInfo *model.SysSubSystemInfo) (*model.SubSystemAuthenticationToken, error) {
	currTime := time.Now()
	currTimeStr := formatDate(currTime)
	if currTimeStr != subSystem.CreateDate {
		return nil, exterror.NewBadCredentialsError(MsgBadCredential)
	}

	if len(subSystem.Nonce) != 10 {
		return nil, exterror.NewBadCredentialsError(MsgBadCredential)
	}

	length, err := strconv.Atoi(subSystem.Nonce[:8])
	if err != nil {
		return nil, err
	}
	//int len = Integer.parseInt(subSystem.getNonce().substring(8));
	if length != len(subSystem.SystemCode) {
		return nil, exterror.NewBadCredentialsError(MsgBadCredential)
	}

	return createSuccessAuthentication(subSystemInfo, subSystem), nil
}

func createSuccessAuthentication(retrievedSubSystemInfo *model.SysSubSystemInfo, subSystem *model.SubSystemTokenDto) *model.SubSystemAuthenticationToken {
	returnAuthToken := &model.SubSystemAuthenticationToken{
		Principal:   subSystem.SystemCode,
		Credentials: subSystem.SystemCode,
		Nonce:       subSystem.Nonce,
		Authorities: retrievedSubSystemInfo.Authorities,
	}

	return returnAuthToken

}

func formatDate(dateTime time.Time) string {
	layout := "20060102" // yyyyMMdd

	return dateTime.Format(layout)
}

func validateStrictPermission(currUser *model.AuthenticatedUser) error {

	if currUser == nil {
		return exterror.NewAuthServerError("Lack of permission.")
	}

	userRoles := currUser.GrantedAuthorities
	if len(userRoles) == 0 {
		return exterror.NewAuthServerError("Lack of permission due to empty roles.")
	}

	if !utils.Contains(userRoles, constant.AuthoritySubsystem) {
		return exterror.NewAuthServerError("Lack of permission due to no sub-system authority.")
	}

	if utils.EqualsIgnoreCase(constant.SubSystemNameSysPlatform, currUser.Username) {
		return nil
	}

	log.Info(nil, log.LOGGER_APP, fmt.Sprintf("current username is : %v", currUser.Username))
	return exterror.NewAuthServerError("Lack of permission due to not platform identity.")
}

func (SubSystemManagementService) RegisterSubSystem(subSystemDto *model.SimpleSubSystemDto, currUser *model.AuthenticatedUser) (*model.SimpleSubSystemDto, error) {

	err := validateStrictPermission(currUser)
	if err != nil {
		return nil, err
	}

	if len(subSystemDto.SystemCode) == 0 {
		return nil, exterror.Catch(exterror.New().AuthServer3016Error, nil)
		//throw new AuthServerException("3016", "Registering sub-system errors:system code cannot be blank.");
	}

	subSystem, err := db.SubSystemRepositoryInstance.FindOneBySystemCode(subSystemDto.SystemCode)
	if err != nil {
		log.Error(nil, log.LOGGER_APP, "failed to find SubSystem", zap.String("systemCode", subSystemDto.SystemCode),
			zap.Error(err))
		return nil, err
	}

	if subSystem == nil {
		subSystem, err = db.SubSystemRepositoryInstance.FindOneBySystemName(subSystemDto.Name)
		if err != nil {
			log.Error(nil, log.LOGGER_APP, "failed to find subSystem by system name", zap.String("subsystem name", subSystemDto.Name),
				zap.Error(err))
			return nil, err
		}
		//subSystem = subSystemRepository.findOneBySystemName(subSystemDto.getName());
	}

	if subSystem != nil {
		log.Debug(nil, log.LOGGER_APP, fmt.Sprintf("such sub-system already exists,system code %s", subSystemDto.SystemCode))
		return convertToSimpleSubSystemDto(subSystem), nil
	}

	log.Info(nil, log.LOGGER_APP, "About to create a new sub system:", log.JsonObj("subSystemDto", subSystemDto))

	keyPair, err := utils.InitAsymmetricKeyPair()
	if err != nil {
		log.Error(nil, log.LOGGER_APP, "failed to init asymmetric key paire", zap.Error(err))
		return nil, err
	}

	subSystem = &model.SysSubSystemEntity{
		Id: utils.Uuid(),
		//TODO
		//CreatedBy:   currUser.Username,
		Description: subSystemDto.Description,
		ApiKey:      keyPair.PrivateKey,
		PubApiKey:   keyPair.PublicKey,
		Name:        subSystemDto.Name,
		SystemCode:  subSystemDto.SystemCode,
		IsActive:    true,
		IsBlocked:   false,
	}

	result, err := db.Engine.Insert(subSystem)
	if err != nil || result == 0 {
		if err != nil {
			log.Error(nil, log.LOGGER_APP, "failed to insert subsystem", log.JsonObj("subsystem", subSystem))
		}
		return nil, errors.New("failed to insert subsystem")
	}

	return convertToSimpleSubSystemDto(subSystem), nil
}

func convertToSimpleSubSystemDto(subSystem *model.SysSubSystemEntity) *model.SimpleSubSystemDto {
	return &model.SimpleSubSystemDto{
		ID:          subSystem.Id,
		Active:      subSystem.IsActive,
		Blocked:     subSystem.IsBlocked,
		Description: subSystem.Description,
		Name:        subSystem.Name,
		SystemCode:  subSystem.SystemCode,
		APIKey:      subSystem.ApiKey,
		PubKey:      subSystem.PubApiKey,
	}
}

func (SubSystemManagementService) RetrieveSubSystemApikey(systemCode string) (*model.SimpleSubSystemDto, error) {
	result := &model.SimpleSubSystemDto{}
	if len(systemCode) == 0 {
		return result, nil
	}

	subSystem, err := db.SubSystemRepositoryInstance.FindOneBySystemCode(systemCode)
	if err != nil {
		log.Error(nil, log.LOGGER_APP, "failed to find SubSystem by systemcode", zap.String("systemcode", systemCode),
			zap.Error(err))
		return nil, err
	}

	if subSystem == nil {
		return result, nil
	}

	return &model.SimpleSubSystemDto{
		ID:          subSystem.Id,
		Active:      subSystem.IsActive,
		Blocked:     subSystem.IsBlocked,
		Description: subSystem.Description,
		Name:        subSystem.Name,
		SystemCode:  subSystem.SystemCode,
		APIKey:      subSystem.ApiKey,
	}, nil
}

func (SubSystemManagementService) RetrieveSubSystemByName(name string, currUser *model.AuthenticatedUser) (*model.SimpleSubSystemDto, error) {
	if err := validateStrictPermission(currUser); err != nil {
		return nil, err
	}
	subSystem, err := db.SubSystemRepositoryInstance.FindOneBySystemName(name)
	if err != nil {
		log.Error(nil, log.LOGGER_APP, "failed to find Subsystem by system name", zap.String("system name", name),
			zap.Error(err))
		return nil, err
	}

	if subSystem == nil {
		return nil, nil
	}

	return convertToSimpleSubSystemDto(subSystem), nil
}

func (SubSystemManagementService) RetrieveAllSubSystems(currUser *model.AuthenticatedUser) ([]*model.SimpleSubSystemDto, error) {
	if err := validateStrictPermission(currUser); err != nil {
		return nil, err
	}
	subSystems := make([]*model.SysSubSystemEntity, 0)
	if err := db.Engine.Find(&subSystems); err != nil {
		log.Error(nil, log.LOGGER_APP, "failed to find subsystem", zap.Error(err))
		return nil, err
	}

	result := make([]*model.SimpleSubSystemDto, 0)
	if len(subSystems) == 0 {
		return nil, nil
	}

	for _, subSystem := range subSystems {
		d := convertToSimpleSubSystemDto(subSystem)
		result = append(result, d)
	}
	return result, nil
}
